Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp

```diff
@@ -422,4 +458,77 @@ void GraphicsAdapter::delete_resource(ResourceID resource_id)
     VERIFY(response.type == static_cast<u32>(Protocol::CommandType::VIRTIO_GPU_RESP_OK_NODATA));
 }
 
+void GraphicsAdapter::initialize_3d_device()
+{
+    if (m_has_virgl_support) {
+        MutexLocker locker(m_operation_lock);
+        m_3d_device = MUST(DeviceManagement::try_create_device<VirtIOGPU::GPU3DDevice>(*this));
+    }
+}
+
+ContextID GraphicsAdapter::create_context()
+{
+    VERIFY(m_operation_lock.is_locked());
+    auto ctx_id = allocate_context_id();
+    auto writer = create_scratchspace_writer();
+    auto& request = writer.append_structure<Protocol::ContextCreate>();
+    auto& response = writer.append_structure<Protocol::ControlHeader>();
+
+    constexpr char const* region_name = "Serenity VirGL3D Context";
+    populate_virtio_gpu_request_header(request.header, Protocol::CommandType::VIRTIO_GPU_CMD_CTX_CREATE, VIRTIO_GPU_FLAG_FENCE);
+    request.header.context_id = ctx_id.value();
+    request.name_length = strlen(region_name);
+    for (auto& v : request.debug_name)
+        v = 0;
```

Comment:
Is this a `memset(request.debug_name.data(), 0, 64)`?

Reply:
True, will fix in next revision.